src/Agent.Listener/Configuration/NativeWindowsServiceHelper.cs

```diff
@@ -808,28 +921,123 @@ public LsaPolicy()
                 };
 
                 IntPtr handle = IntPtr.Zero;
-                uint result = LsaOpenPolicy(ref system, ref attrib, LSA_POLICY_ALL_ACCESS, out handle);
-                if (result != 0 || handle == IntPtr.Zero)
+                uint hr = LsaOpenPolicy(ref system, ref attrib, (uint)access, out handle);
+                if (hr != 0 || handle == IntPtr.Zero)
                 {
-                    throw new Exception(StringUtil.Loc("OperationFailed", nameof(LsaOpenPolicy), result));
+                    throw new Exception(StringUtil.Loc("OperationFailed", nameof(LsaOpenPolicy), hr));
                 }
 
                 Handle = handle;
             }
 
+            public void SetSecretData(string key, string value)
+            {
+                LSA_UNICODE_STRING secretData = new LSA_UNICODE_STRING();
+                LSA_UNICODE_STRING secretName = new LSA_UNICODE_STRING();
+
+                secretName.Buffer = Marshal.StringToHGlobalUni(key);                
+                
+                var charSize = sizeof(char);
+
+                secretName.Length = (UInt16)(key.Length * charSize);
+                secretName.MaximumLength = (UInt16)((key.Length + 1) * charSize);
+
+                if (value != null && value.Length > 0)
+                {
+                    // Create data and key
+                    secretData.Buffer = Marshal.StringToHGlobalUni(value);
+                    secretData.Length = (UInt16)(value.Length * charSize);
+                    secretData.MaximumLength = (UInt16)((value.Length + 1) * charSize);
+                }
+                else
+                {
+                    // Delete data and key
+                    secretData.Buffer = IntPtr.Zero;
+                    secretData.Length = 0;
+                    secretData.MaximumLength = 0;
+                }
+
+                uint result = LsaStorePrivateData(Handle, ref secretName, ref secretData);
+                uint winErrorCode = LsaNtStatusToWinError(result);
+                if (winErrorCode != 0)
+                {
+                    throw new Exception(StringUtil.Loc("OperationFailed", nameof(LsaNtStatusToWinError), winErrorCode));
+                }
+            }
+
             void IDisposable.Dispose()
             {
                 // We will ignore LsaClose error
                 LsaClose(Handle);
                 GC.SuppressFinalize(this);
             }
+
+            internal static string DefaultPassword = "DefaultPassword";
         }
 
-        // Declaration of external pinvoke functions
-        private static readonly uint LSA_POLICY_ALL_ACCESS = 0x1FFF;
-        private static readonly string s_logonAsServiceName = "SeServiceLogonRight";
+        // Class to represent a failure action which consists of a recovery
+        // action type and an action delay
+        private class FailureAction
```

Comment:
Can we try not change the order of this, so we can make the diff view more clear, since github diff experience is so bad. :(